### ISSUE TITLE
[Veinminer] Smarter search

### ISSUE THREAD

**Issue by Derpius** [TAServers/tas-mineclone2#17]
## Context

The current search algorithm is perfectly valid, but produces results that can look “wrong” to players. This is because the offsets are hardcoded world directions.

## Proposed solution

- Treat the offsets as local vectors in a broken block face’s TBN coordinate system
- When a user breaks a block, get the normal of the face that the user mined (this should be possible, but might need a raycast instead of being returns)
- To calculate the bitangent (and by extension the tangent), use the following rules:
  - If the face is vertical, the bitangent is always up
  - If the face is horizontal, the bitangent is the up vector of the player’s camera, with y set to 0 and normalised, and then rounded to the nearest orthogonal direction
- Use these vectors to transform the local offsets to world vectors

Once the above work is complete, we’ll likely also need to order the offset tables better to produce a more hand-crafted result

**Comment by yogwoggf** [comment]
... Did some.. investigation and it appears that the only way to figure out the normal of the currently faced block is to raycast?? But here's the thing, apparently, errors "build up" in multiplayer games. Here's another whammy: there's no way to get the camera position (eye pos equivalent in gmod) in Minetest either! All we can do is get the camera's current direction, and the patched callback I'm using doesn't provide a "pointed_thing." (but the information we need is not present in any callback, only via raycasts).

Not sure what to do. @Derpius

**Comment by Derpius** [comment]
Raycasts is what I expected we needed anyway, and we only need the camera’s direction.

If we can’t get the camera’s position, then raycasts must always come from the eye otherwise how are stairs and logs etc. placed?

**Comment by yogwoggf** [comment]
> Raycasts is what I expected we needed anyway, and we only need the camera’s direction.
> 
> If we can’t get the camera’s position, then raycasts must always come from the eye otherwise how are stairs and logs etc. placed?

No, the raycast API is sort of like GMod, you give it a starting position and a ending position. Block placing is handled by the engine.
